database/database.rs

```diff
@@ -185,37 +213,38 @@ impl Database<WALClient> {
         let _schema_write_lock = self.schema_txn_lock.write().unwrap();
 
         match Arc::get_mut(&mut self.storage) {
-            None => {
-                return Err(DatabaseResetError::StorageInUse {});
-            }
+            None => return Err(DatabaseResetError::StorageInUse {}),
             Some(storage) => storage.reset().map_err(|err| CorruptionStorageReset { source: err })?,
         }
         match Arc::get_mut(&mut self.definition_key_generator) {
-            None => {
-                return Err(CorruptionDefinitionKeyGeneratorInUse {});
-            }
+            None => return Err(CorruptionDefinitionKeyGeneratorInUse {}),
             Some(definition_key_generator) => definition_key_generator.reset(),
         }
         match Arc::get_mut(&mut self.type_vertex_generator) {
-            None => {
-                return Err(CorruptionTypeVertexGeneratorInUse {});
-            }
+            None => return Err(CorruptionTypeVertexGeneratorInUse {}),
             Some(type_vertex_generator) => type_vertex_generator.reset(),
         }
         match Arc::get_mut(&mut self.thing_vertex_generator) {
-            None => {
-                return Err(TypeVertexGeneratorInUse {});
-            }
+            None => return Err(TypeVertexGeneratorInUse {}),
             Some(thing_vertex_generator) => thing_vertex_generator.reset(),
         }
 
-        let schema = Arc::get_mut(&mut *locked_schema).unwrap();
-        schema.thing_statistics.reset(self.storage.read_watermark());
+        let thing_statistics = Arc::get_mut(&mut locked_schema.thing_statistics).unwrap();
+        thing_statistics.reset(self.storage.read_watermark());
 
         Ok(())
     }
 }
 
+fn update_statistics_action(storage: Arc<MVCCStorage<WALClient>>, schema: Arc<RwLock<Schema>>) -> impl Fn() {
+    move || {
+        let mut schema_lock = schema.write().unwrap();
+        let mut thing_statistics = (*schema_lock.thing_statistics).clone();
+        thing_statistics.may_synchronise(&storage).ok();
+        schema_lock.thing_statistics = Arc::new(thing_statistics);
+    }
+}
```

Comment:
This will block read transactions from opening - if there are many records in the WAL we have to catch up, even 10-60 seconds could make this operation non-trivial.
I suggest that we two the locking in two stages:
1. obtain a read-lock, get a clone of the statistics and synchronise it
2. upgrade to a write lock, and then replace the statistics object.

nb. can we atomically swap the read lock for a write lock? I think it's probably required for safety - eg. if a schema txn opens in between, then commits, we get into a skewed state

Reply:
Also let's call this function `update_statistics_runnable` or `_fn` or something? Action is a bit odd IMO